File: Python/BinaryTreeTraversal.py
from collections import deque
import heapq

class TreeNode:
    def __init__(self, value):
        self.val = value
        self.left = None
        self.right = None
    
def preOrder(node: TreeNode):
        if node is None:
             return
        print(node.value, end=" ")
        preOrder(node.left)
        preOrder(node.right)

def inOrder(node: TreeNode):
        if node is None:
             return
        inOrder(node.left)
        print(node.value, end=" ")
        inOrder(node.right)
        

def postOrder(node: TreeNode):
        if node is None:
             return

        postOrder(node.left)
        postOrder(node.right)
        print(node.value, end=" ")

def levelOrder(root: TreeNode) -> list[list[int]]:
    #Problem #102. Binary Tree Level Order Traversal - Medium - I learned from a youtuber Greg Hogg

    if root is None:
        return
    
    q = deque()
    q.append(root)
    ans = []

    while q:
        level = []
        for i in range(len(q)):
            node = q.popleft()
            level.append(node.val)
            if node.left: q.append(node.left)
            if node.right: q.append(node.right)
        ans.append(level)
        
    return ans

def binaryTreePaths(root: TreeNode) -> list[str]:
    #Problem #257 Binary Tree Paths - Easy
    
    rootToLeaf = []
    pathToLeaf = []
    
    def preOrderTraversal(node: TreeNode):     
        if node == None:
            return
        
        if node.left == None and node.right == None:
            pathToLeaf.append(str(node.val))
            rootToLeaf.append("".join(pathToLeaf))
            pathToLeaf.pop()
            return
        
        pathToLeaf.append(f'{node.val}->')
        
        preOrderTraversal(node.left)
        preOrderTraversal(node.right)
        
        pathToLeaf.pop()
    
    preOrderTraversal(root)
    
    return rootToLeaf    
         
def kthSmallest(root: TreeNode, k: int) -> int:
    #Problem #230 Kth Smallest Element in BST - Medium
    
    inOrder = []
    
    def inOrderTraversal(node: TreeNode):
        if node == None:
            return
        
        inOrderTraversal(node.left)
        inOrder.append(node.val)
        inOrderTraversal(node.right)
            
    inOrderTraversal(root)
    
    return inOrder[k - 1]

def maxDepth(root: TreeNode) -> int:
    #Problem #104 Maximum Depth of Binary Tree - Easy
     
    if root == None:
        return 0
    
    depthCount = 0
    
    nodeQueue = deque([root])
    
    while nodeQueue:
        depthCount += 1
        for index in range(len(nodeQueue)):
            node = nodeQueue.popleft()
            if node.left != None:
                nodeQueue.append(node.left)
            
            if node.right != None:
                nodeQueue.append(node.right)
            
    return depthCount

def widthOfBinaryTree(root: TreeNode) -> int:
    #Problem #662 Maximum Width of Binary Tree - Medium - Solution Concept by YouTube Channel Timothy H Chang - Understanding the Solution
    
    nodeQueueIndex =  deque([(root, 1)])
    maxWidth = 0
    
    while nodeQueueIndex:
        leftMostNode, lefttIndex = nodeQueueIndex[0]
        rightMostNode, rightIndex = nodeQueueIndex[len(nodeQueueIndex) - 1]
        
        maxWidth = max(maxWidth, rightIndex - lefttIndex + 1)
        
        for index in range(len(nodeQueueIndex)):
            node, indexNode = nodeQueueIndex.popleft()
            if node.left != None:
                nodeQueueIndex.append((node.left, 2 * indexNode))
            if node.right != None:
                nodeQueueIndex.append((node.right, 2 * indexNode + 1))
    
    return maxWidth

def maxPathSum(root: TreeNode) -> int:
    #Problem #124 Binary Tree Maximum Path Sum - Hard - Solution from a Submitted Code in LeetCode 
    #Understanding the Solution

    maxSum = [float('-inf')]
    
    def maxPathDFS(node: TreeNode):
        if node == None:
            return 0
        
        if node.left == None and node.right == None:
            maxSum[0] = max(maxSum[0], node.val)
            return node.val
        
        leftNodeValue = max(maxPathDFS(node.left), 0)
        rightNodeValue = max(maxPathDFS(node.right), 0)
        
        maxSum[0] = max(maxSum[0], leftNodeValue + node.val + rightNodeValue)
        
        return max(leftNodeValue, rightNodeValue) + node.val
    
    maxPathDFS(root)
    
    return maxSum[0]
    
def levelOrderBottom(root: TreeNode) -> list[list[int]]:
    #Problem #107 Binary Tree Level Order Traversal II - Medium
    
    if root == None:
        return []
    
    nodeQueue = deque([root])
    bottomToTopLevelOrder = []
    levelValues = [root.val]
    
    while nodeQueue:
        bottomToTopLevelOrder.append(levelValues)
        levelValues = []
        for index in range(len(nodeQueue)):
            node = nodeQueue.popleft()
            if node.left != None:
                levelValues.append(node.left.val)
                nodeQueue.append(node.left)
            if node.right != None:
                levelValues.append(node.right.val)
                nodeQueue.append(node.right)
    
    bottomToTopLevelOrder.reverse()
    
    return bottomToTopLevelOrder

if __name__ == "__main__":
   node1 = TreeNode(1)
   node2 = TreeNode(2)
   node3 = TreeNode(3)
   node3Duplicate = TreeNode(3)
   node4 = TreeNode(4)
   node5 = TreeNode(5)
   node8 = TreeNode(8)
   node9 = TreeNode(9)
   node20 = TreeNode(20)
   node15 = TreeNode(15)
   node7 = TreeNode(7)
   node6 = TreeNode(6)
   nodeNegative10 = TreeNode(-10)
   
   nodeNegative10.left = node9
   nodeNegative10.right = node20
   node20.left = node15
   node20.right = node7

   root = nodeNegative10
   
   print(maxPathSum(root))

File: C++/C++/BinaryTreeTraversal.cpp
//
//  BinaryTreeTraversal.cpp
//  C++
//
//  Created by Richmond Laureta on 6/23/25.
//

#include "Header.h"

TreeNode::TreeNode(int val) : val(val), left(nullptr), right(nullptr){}

void preOrder(TreeNode *node)
{
    if(node == nullptr)
    {
        return;
    }
    
    cout << node->val << " ";
    preOrder(node->left);
    preOrder(node->right);
}

void inOrder(TreeNode *node)
{
    if(node == nullptr)
    {
        return;
    }
    
    inOrder(node->left);
    cout << node->val << " ";
    inOrder(node->right);
}


void postOrder(TreeNode *node)
{
    if(node == nullptr)
    {
        return;
    }
    
    postOrder(node->left);
    postOrder(node->right);
    cout << node->val << " ";
}

vector<vector<int>> levelOrder(TreeNode *node)
{
//Problem #102 Binary Tree Level Order Traversal
    
    if(node == nullptr)
    {
        return {};
    }
    vector<vector<int>> myList;
    queue<TreeNode*> myQueue;
    
    myQueue.push(node);
    
    while(!myQueue.empty())
    {
        vector<int> level;
        size_t length = myQueue.size();
        for(size_t i = 0; i < length; i++)
        {
            level.push_back(myQueue.front()->val);
            TreeNode *nodeCopy = myQueue.front();
            myQueue.pop();
            
            if(nodeCopy->left != nullptr) {myQueue.push(nodeCopy->left);}
            if(nodeCopy->right != nullptr) {myQueue.push(nodeCopy->right);}
        }
        length = myQueue.size();
        myList.push_back(level);
    }
    return myList;
}

void btpPreorderTraversal(TreeNode *node, vector<string> &rootToLeaf, vector<string> &pathToLeaf)
{
    //Problem #257 Binary Tree Paths - Easy
    
    if(node == nullptr) return;
    
    if(node->left == nullptr and node->right == nullptr)
    {
        pathToLeaf.push_back(to_string(node->val));
        rootToLeaf.push_back(accumulate(pathToLeaf.begin(), pathToLeaf.end(), string("")));
        pathToLeaf.pop_back();
        return;
    }
    
    pathToLeaf.push_back(to_string(node->val) + "->");
    
    btpPreorderTraversal(node->left, rootToLeaf, pathToLeaf);
    btpPreorderTraversal(node->right, rootToLeaf, pathToLeaf);
    
    pathToLeaf.pop_back();
}

vector<string> binaryTreePaths(TreeNode *root)
{
    //Problem #257 Binary Tree Paths - Easy
    
    vector<string> rootToLeaf = {};
    vector<string> pathToLeaf = {};
    
    btpPreorderTraversal(root, rootToLeaf, pathToLeaf);
    
    return rootToLeaf;
}

void inOrderTraversal(TreeNode *node, vector<int> &ascendingArray)
{
    //Problem #230 Kth Smallest Element in BST
    
    if(node == nullptr) return;
    
    inOrderTraversal(node->left, ascendingArray);
    ascendingArray.push_back(node->val);
    inOrderTraversal(node->right, ascendingArray);
}

int kthSmallest(TreeNode *root, int k)
{
    //Problem #230 Kth Smallest Element in BST
    
    vector<int> ascendingArray;
    inOrderTraversal(root, ascendingArray);
    return ascendingArray[k - 1];
}

int maxDepth(TreeNode* root)
{
    //Problem #104 Maximum Depth of Binary Tree - Easy
    
    if(root == nullptr) return 0;
    
    int depthCount = 0;
    
    deque<TreeNode*> nodeQueue = {root};
    
    while((int)nodeQueue.size() > 0)
    {
        depthCount += 1;
        int nodeQueueSize = (int)nodeQueue.size();
        for(int index = 0; index < nodeQueueSize; index++)
        {
            TreeNode *node = nodeQueue.front();
            nodeQueue.pop_front();
            
            if(node->left != nullptr) nodeQueue.push_back(node->left);
            
            if(node->right != nullptr) nodeQueue.push_back(node->right);
        }
    }
    
    return depthCount;
}

int widthOfBinaryTree(TreeNode *root)
{
    //Problem #662 Maximum Width of Binary Tree - Medium - Solution Concept by YouTube Channel Timoth H Chang
    
    deque<pair<TreeNode*, int>> nodeQueueIndex = {{root, 1}};
    
    int maxWidth = 0;
    
    while((int)nodeQueueIndex.size() > 0)
    {
        int leftIndex = nodeQueueIndex[0].second;
        int rightIndex = nodeQueueIndex[(int)nodeQueueIndex.size() - 1].second;
        
        maxWidth = max(maxWidth, rightIndex - leftIndex + 1);
        
        int nodeQueueIndexSize = (int)nodeQueueIndex.size();
        
        for(int index = 0; index < nodeQueueIndexSize; index++)
        {
            TreeNode* node = nodeQueueIndex.front().first;
            int nodeIndex = nodeQueueIndex.front().second;
            nodeQueueIndex.pop_front();
            if(node->left != nullptr) nodeQueueIndex.push_back({node->left, 2 * nodeIndex});
            if(node->right != nullptr) nodeQueueIndex.push_back({node->right, 2 * nodeIndex + 1});
        }
    }
    
    return maxWidth;
}

int depthFirstSearchSum(TreeNode *node, int *maxSum)
{
    //Problem #124 Binary Tree Maximum Path Sum - Solution Concept by NeetCode - Understanding the Solution
    
    if(node == nullptr)
    {
        return 0;
    }
    
    int leftSum = depthFirstSearchSum(node->left, maxSum);
    int rightSum = depthFirstSearchSum(node->right, maxSum);
    
    leftSum = max(leftSum, 0);
    rightSum = max(rightSum, 0);
    
    *maxSum = max(*maxSum, node->val + leftSum + rightSum);
    
    return node->val + max(leftSum, rightSum);
}

int maxPathSum(TreeNode *root)
{
    //Problem #124 Binary Tree Maximum Path Sum - Solution Concept by NeetCode - Understanding the Solution
    
    int maxSum = INT_MIN;
    depthFirstSearchSum(root, &maxSum);
    return maxSum;
}

vector<vector<int>> levelOrderBottom(TreeNode *root)
{
    //Problem #107 Binary Tree Level Order Traversal II - Medium
    
    if(root == nullptr) return {};
    
    deque<TreeNode*> nodeQueue = {root};
    vector<vector<int>> bottomTopLevelOrderVector = {};
    vector<int> levelValues = {root->val};
    
    while((int)nodeQueue.size() > 0)
    {
        bottomTopLevelOrderVector.push_back(levelValues);
        levelValues = {};
        int nodeQueueSize = (int) nodeQueue.size();
        
        for(int index = 0; index < nodeQueueSize; index++)
        {
            TreeNode *node = nodeQueue.front();
            nodeQueue.pop_front();
            
            if(node->left != nullptr)
            {
                levelValues.push_back(node->left->val);
                nodeQueue.push_back(node->left);
            }
            
            if(node->right != nullptr)
            {
                levelValues.push_back(node->right->val);
                nodeQueue.push_back(node->right);
            }
        }
    }
    
    reverse(bottomTopLevelOrderVector.begin(), bottomTopLevelOrderVector.end());
    
    return bottomTopLevelOrderVector;
}
